src/shogun/features/CombinedDotFeatures.cpp

```diff
@@ -243,42 +243,44 @@ void CombinedDotFeatures::free_feature_iterator(void* iterator)
 
 std::shared_ptr<DotFeatures> CombinedDotFeatures::get_feature_obj(int32_t idx) const
 {
-	return feature_array->get_element<DotFeatures>(idx);
+	return feature_array[idx]->as<DotFeatures>();
 }
 
 bool CombinedDotFeatures::insert_feature_obj(std::shared_ptr<DotFeatures> obj, int32_t idx)
 {
 	ASSERT(obj)
-	bool result=feature_array->insert_element(obj, idx);
-	auto iterator = feature_weights.begin() + idx;
-	feature_weights.insert(iterator, initial_weight);
+	feature_array.insert(feature_array.begin() + idx, obj);
+	feature_weights.insert(feature_weights.begin() + idx, initial_weight);
 	update_dim_feature_space_and_num_vec();
-	return result;
+	return true;
 }
 
 bool CombinedDotFeatures::append_feature_obj(std::shared_ptr<DotFeatures> obj)
 {
 	ASSERT(obj)
 	int n = get_num_feature_obj();
-	feature_array->push_back(obj);
+	feature_array.push_back(obj);
```

Comment:
in these situations it might be worth doing std::move. You copy the shared_ptr anyway because you pass by value, so just move that copy in the push_back, no?

Reply:
yup